netkan/netkan/webhooks/github_mirror.py

```diff
@@ -0,0 +1,315 @@
+
+    def download_content_type(self):
+        return self._raw['download_content_type']
+
+    def download_size(self):
+        return self._raw['download_size']
+
+    def download(self):
+        return self._raw['download']
+
+    def mirrored(self, ia):
+        results = ia.search_items(self.mirror_item())
+        return True if results else False
+
+
+def ia_overloaded():
+    return requests.get(
+        'https://s3.us.archive.org/?check_limit=1&accesskey={}'.format(
+            current_app.config['ia_access']
+        )
+    ).status_code == 503
+
+
+def mirror(ckans):
+    if not ckans:
+        current_app.logger.info('No ckans found')
+        return
+    current_app.logger.info('Setting IA credentials')
+    if ia_overloaded():
+        current_app.logger.info('The Internet Archive is overloaded, try again later')
+        return
+    current_app.logger.info('Not overloaded')
+    # Prepare archive.org connection
+    ia = internetarchive.get_session(config = {
+        's3': {
+            'access': current_app.config['ia_access'],
+            'secret': current_app.config['ia_secret'],
+        }
+    })
+    for ckan in ckans:
+        current_app.logger.info(f'Handling {ckan.mirror_item()}')
+        if not ckan.can_mirror():
+            current_app.logger.info(f'Ckan {ckan.mirror_item()} cannot be mirrored')
+        elif ckan.mirrored(ia):
+            current_app.logger.info(f'Ckan {ckan.mirror_item()} already mirrored')
+        else:
+            current_app.logger.info('Creating temp file')
+            with tempfile.NamedTemporaryFile() as tmp:
+                current_app.logger.info(f'Downloading {ckan.download()}')
+                urllib.request.urlretrieve(ckan.download(), tmp.name)
+                current_app.logger.info(f'Downloaded {ckan.mirror_item()} to {tmp.name}')
+                if ckan.download_sha256() != hashlib.sha256(tmp.file.read()).hexdigest().upper():
+                    current_app.logger.info(f'Hash mismatch')
+                else:
+                    lic_urls = ckan.license_urls()
+                    item = internetarchive.Item(ia, ckan.mirror_item())
+                    current_app.logger.info(f'Uploading')
+                    item.upload_file(tmp.name, ckan.mirror_item(), {
+                        'title':       ckan.mirror_title(),
+                        'description': ckan.mirror_description(),
+                        'creator':     ckan.authors(),
+                        'collection':  current_app.config['ia_collection'],
+                        'subject':     'ksp; kerbal space program; mod',
+                        'mediatype':   'software',
+                        **({ 'licenseurl': lic_urls } if lic_urls else {}),
+                    }, {
+                        'Content-Type':           ckan.download_content_type(),
+                        'Content-Length':         ckan.download_size(),
+                        'x-amz-auto-make-bucket': 1,
+                    })
```

Comment:
This is asking a lot of the flask app, one of the things that I found awkward with the previous architecture was that each piece was being asked to do too much, as we didn't have any other way to achieve it. 

I have some ideas here, but none of this work is wasted. It'll be easy enough to move it around as needed.